### PR TITLE
[8.x] Track exceptions and display them on failed status checks for dx

### PR DESCRIPTION
**Pull request by taylorotwell**
This feature is inspired by feedback from @jessarcher and my own poor experience with testing endpoints that throw unexpected exceptions.

This problem typically arises when asserting against the status of a test response and receiving a failed assertions indicating the expected status code did not match the actual status code of `500`. You must then examine your logs to hopefully determine what the exception message was. Or, you can call `$this->withoutExceptionHandling()->get(...)` to disable exception handling for the test.

Ideally, Laravel would just show you the exception. This PR takes care of that by tracking exceptions while running unit tests and then displaying the exception failed status checks where the actual status code is `500`.

image

image

### CODE REVIEW

File: src/Illuminate/Foundation/Testing/Concerns/MakesHttpRequests.php [hunk 1]
```diff
@@ -641,6 +642,10 @@ protected function followRedirects($response)
      */
     protected function createTestResponse($response)
     {
-        return TestResponse::fromBaseResponse($response);
+        return tap(TestResponse::fromBaseResponse($response), function ($response) {
+            $response->withExceptions($this->app->bound(LoggedExceptionCollection::class)
+                            ? $this->app->make(LoggedExceptionCollection::class)
+                            : new LoggedExceptionCollection);
```
Comment (driesvints) [inline src/Illuminate/Foundation/Testing/Concerns/MakesHttpRequests.php:648]: It took me a sec to digest this. Try:

```suggestion
            $response->withExceptions(
                $this->app->bound(LoggedExceptionCollection::class)
                    ? $this->app->make(LoggedExceptionCollection::class)
                    : new LoggedExceptionCollection
            );
```

**Review by driesvints — APPROVED** [review]

**Comment by negoziator** [comment]
It's uncountable how many times I've wanted this.
Great job! 🎉

**Comment by deleugpn** [comment]
👌

**Comment by rachids** [comment]
To all those `$this->withoutExceptionHandling()` that hasn't been removed and managed to pass the code review: **you're in danger.**

**Review by jessarcher — APPROVED** [review]
This is a brilliant DX improvement and will save a lot of time and frustration! Thank you :grin:

**Comment by jessarcher** [comment]
In the spirit of removing the need to add temporary debugging code, I wonder whether there is also an opportunity to improve the output on an unexpected validation exception.

Currently I get this a fair bit:

image

Even without exception handling (which this PR removes the need for :grin:), the output is no more helpful:

image

So I end up needing to add a temporary `$response->dump()` to see the actual validation errors.

If this sounds useful, I'm happy to have a go once this PR is merged. I'm thinking if it's a JSON request then it should show the errors as JSON, otherwise perhaps just a `var_export()`.

**Comment by taylorotwell** [comment]
@jessarcher sounds like a good idea

**Pull request merged**